### PR TITLE
Let container scope string invoke be `named`

### PR DESCRIPTION
**Pull request by eskatos**
See #907

### CODE REVIEW

File: subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 1]
```diff
@@ -66,28 +66,28 @@ class NamedDomainObjectContainerScope<T : Any>(
         polymorphicDomainObjectContainer().containerWithType(type)
 
     /**
-     * @see [NamedDomainObjectContainer.maybeCreate]
+     * @see [NamedDomainObjectContainer.named]
      */
-    inline operator fun String.invoke(configuration: T.() -> Unit): T =
-        this().apply(configuration)
+    operator fun String.invoke(configuration: T.() -> Unit): DomainObjectProvider<T> =
+        this().also { it.configure(configuration) }
```
Comment (bamboo) [inline subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: ⭕️ Why not `this().apply { configure(configuration) }`?

File: subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt [hunk 2]
```diff
@@ -66,28 +66,28 @@ class NamedDomainObjectContainerScope<T : Any>(
...
+        this(type).also { it.configure(configuration) }
 
     /**
-     * @see [PolymorphicDomainObjectContainer.maybeCreate]
+     * @see [PolymorphicDomainObjectContainer.named]
      */
-    operator fun <U : T> String.invoke(type: KClass<U>): U =
-        polymorphicDomainObjectContainer().maybeCreate(this, type.java)
+    operator fun <U : T> String.invoke(type: KClass<U>): DomainObjectProvider<U> =
+        polymorphicDomainObjectContainer().withType(type.java).named(this)
```
Comment (bamboo) [inline subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: :x: The semantics should be that the caller is asserting the type of the named element, not querying elements of the specific type, similar to [`NamedDomainobjectCollection.getByName<T>(String)`](https://github.com/gradle/kotlin-dsl/blob/448edf838543abb886968bab59e9bc28f1c6111b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensions.kt#L60).

Comment (eskatos) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: The assertion would need to be lazy to not realize the container element, as a configuration block maybe. Then the assertion happens sometime in the future, when/if the element is realized.

Comment (eskatos) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: I agree the semantics are different. If the targeted element cannot be found (by type and name) you'll get a "not found" exception. What you'd not get is the detail about an element found with the requested name but of a different type.

Comment (bamboo) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: Yes, the assertion should happen during configuration.

Comment (eskatos) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: Then it requires an eager cast to return the requested provider type that may fail before we have a chance to provide a helpful error:
```kotlin
operator fun <U : T> String.invoke(type: KClass<U>): DomainObjectProvider<U> =
    polymorphicDomainObjectContainer().named(this).apply {
        configure {
            it as? U
                ?: illegalElementType(container, this@invoke, type, it::class)
        }
    } as DomainObjectProvider<U>
```

Or did I miss something?

Comment (bamboo) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: I was thinking on something like:

```kotlin
    /**
     * @see [PolymorphicDomainObjectContainer.named]
     */
    operator fun <U : T> String.invoke(type: KClass<U>, configuration: U.() -> Unit): DomainObjectProvider<U> =
        uncheckedCast(
            polymorphicDomainObjectContainer().named(this).apply {
                configure {
                    configuration(
                        type.safeCast(it)
                            ?: throw illegalElementType(container, name, type, it::class)
                    )
                }
            })
```

Comment (eskatos) [reply subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt]: Yeah, it's basically the same.
Wouldn't the unchecked cast fail earlier than the type check?

**Pull request merged**